### ISSUE TITLE
napari-imagej can't determine type hint for ArrayImg

### ISSUE THREAD

**Issue by hinerm** [imagej/napari-imagej#56]
Running the `math.add(arg, value)` op results in this exception when generating the widget panel:
![image](https://user-images.githubusercontent.com/1123809/170525976-bdfbb748-e050-4951-952c-d1f9f02fa4af.png)

See https://github.com/imagej/napari-imagej/issues/48#issuecomment-1138870690

**Comment by gselzer** [comment]
Man, that is just a really picky op :confused:

It's an `Inplace` with a [mutable `ArrayImg`](https://github.com/imagej/imagej-ops/blob/9ead9ba0e2ec9604b8c168eb43a9e6162d33f19c/src/main/templates/net/imagej/ops/math/ConstantToArrayImage.vm#L72). PyImageJ couldn't even handle that with a python (e.g. numpy) input image!

Right now, PyImageJ will [convert](https://github.com/imagej/pyimagej/blob/d1cbaaec2e8f9c8b70ffd1a5ff2538b11c0c493d/src/imagej/__init__.py#L900) the python input to an `ImgView`, which is not an `ArrayImg`. 

I'd think we'd be able to convert a numpy `ndarray` into an `ArrayImg` of some sort, but that is no trivial matter. And it would have nothing to do with napari-imagej.

I think we instead would have to "hide" these Ops, but that is also no easy task.

I'm not sure what the best path forward is. What do you think, @hinerm?

**Comment by hinerm** [comment]
>I'm not sure what the best path forward is

I agree that this isn't a problem for napari-imagej. I believe Fiji has the same problem when trying to run ops that require an input `ArrayImg`. So maybe this issue could be moved to ops.

I don't think you need to explicitly hide these Ops in napari-imagej since the problem is consistent across environments.

**Comment by gselzer** [comment]
> I agree that this isn't a problem for napari-imagej. I believe Fiji has the same problem when trying to run ops that require an input `ArrayImg`. So maybe this issue could be moved to ops.

I agree that an issue should be filed in Ops-land, but what issue? There isn't really anything *wrong* with this Op; it is typed on `ArrayImg` because it makes use of `ArrayImg` functionality. The whole point of Ops was that you could write algorithms like this that would only be matched if your inputs were specific enough.

One change we could make would be to write more Ops with the same signature but a broader input object, at a lower priority. This would allow you to get the functionality across the board. **But** it would not fix the problem we are describing here, which is that we expect to run this Op but cannot because its input type is too specific. *Making this change would not avoid the issue, it would just introduce more Ops and probably more confusion*. (As an aside, it sucks to keep filing issues and adding ops in imagej-ops when imagej-ops2 is so close...)

The more impactful change might be what is described in imagej/imagej-ops#643, because adding that typing would give more clarity and could tell us what can be run.

Here's an idea: what if we wrote a `Searcher` implementation that could wrap another `Searcher` with additional predicates? This could allow us to only return `SearchResult`s that e.g. belong to a set of types we know how to resolve.

Here's a draft:

```java
public class SearcherWithPredicates implements Searcher {
    public SearcherWithPredicates(Searcher s, List<? extends Predicate<SearchResult>> predicates) {
        this.s = s;
        this.uberPredicate = predicates.get(0);
        for( i = 1; i < predicates.size(); i++)
            this.uberPredicate = this.uberPredicate.and(predicates.get(i));
    }

    @Override
    public List<SearchResult> search(final String text, final boolean fuzzy) {
        return s.search(text).stream()
          .filter(uberPredicate::test) // Only allow results that satisfy ALL predicates
          .collect(Collections.toList());
    }
}
```

**Comment by ctrueden** [comment]
The core issue here is that ops are not meant to be run directly like this, but rather matched against compatible inputs.

I do not think we should not be exposing individual op signatures. Rather, when you execute a particular op, inputs should be fed to the Ops matcher as per the design, so the right one can be selected by the framework.

Of course, for this to work, we do need to scan all the ops, and aggregate them into supported more general signatures of some kind. Otherwise, there is no way to know what (what types, how many) the user wants to feed to the matcher. It's a sort of chicken-and-egg problem.

**Comment by ctrueden** [comment]
A more complex way to collapse the ops would be:
* Make a list of supported "native" types provided by a particular execution context. In this case, this "execution context" is "napari" so the list of supported "native" types is "distinct Python types compatible with magicgui".
* For each op, loop its inputs, and ask the Python-side conversion framework: can a napari-native type map to this particular op input? If so, which napari-native type?
* Same for output types, but in the other direction.
* For ops where the answer is "no" for any of its parameter types, exclude it from the napari-imagej UI.
* For ops where the answer is yes for all parameter types, produce a "napari-native" type signature for that op, adding it to a set of such signatures to collapse down duplicate ones, since sometimes two or more ops will end up with the same napari-native signature.
* When searching for ops, search the set of napari-native signatures, and display those to the user.

**Comment by gselzer** [comment]
> A more complex way to collapse the ops would be:
> 
> * Make a list of supported "native" types provided by a particular execution context. In this case, this "execution context" is "napari" so the list of supported "native" types is "distinct Python types compatible with magicgui".
> * For each op, loop its inputs, and ask the Python-side conversion framework: can a napari-native type map to this particular op input? If so, which napari-native type?
> * Same for output types, but in the other direction.
> * For ops where the answer is "no" for any of its parameter types, exclude it from the napari-imagej UI.
> * For ops where the answer is yes for all parameter types, produce a "napari-native" type signature for that op, adding it to a set of such signatures to collapse down duplicate ones, since sometimes two or more ops will end up with the same napari-native signature.
> * When searching for ops, search the set of napari-native signatures, and display those to the user.

Yup, this is exactly what I was getting at, and would be easily doable with that `SearcherWithPredicates` class I described above.

**Comment by ctrueden** [comment]
> Which op is the one that gets chosen? You can't use priority.

Yeah, my point is: you aren't choosing specific op implementations. When you present a "napari-native" op signature to the user, they then populate a corresponding magicgui widget, which is then fed *to the ops matcher*, not to any specific implementation. So issues of priority don't matter anymore.

**Comment by gselzer** [comment]
Yeah, that might work out. Will be a bit of work, though.

**Comment by hinerm** [comment]
FWIW I did work isolating ops that operate on images and consolidating parameters in [the op finder](https://imagej.net/plugins/op-finder#running-ops).

**Comment by gselzer** [comment]
> FWIW I did work isolating ops that operate on images and consolidating parameters in [the op finder](https://imagej.net/plugins/op-finder#running-ops).

Thanks @hinerm, I'll probably try to use as much of that as I can. It would be cool if we could migrate some of that to imagej-ops.

**Comment by gselzer** [comment]
While the `OpListing` issue is fixed, this issue now has the same issue as #51. It really sucks that Ops doesn't expose the generics...

@ctrueden what's the best way forward here?

**Comment by gselzer** [comment]
Closing this as duplicate of imagej/imagej-ops#648. You can reproduce in Fiji using the following script, yielding the following error:
```groovy
#@ Img input
#@ OpService ops

import net.imglib2.type.numeric.integer.IntType

ops.math().add(input, new IntType(10))
```
Error:
```
java.lang.ClassCastException: net.imglib2.type.numeric.integer.IntType cannot be cast to net.imglib2.type.numeric.integer.UnsignedByteType
	at net.imglib2.type.numeric.integer.UnsignedByteType.add(UnsignedByteType.java:50)
	at net.imagej.ops.math.ConstantToIIOutputII$Add.compute(ConstantToIIOutputII.java:85)
	at net.imagej.ops.math.ConstantToIIOutputII$Add.compute(ConstantToIIOutputII.java:58)
	at net.imagej.ops.special.hybrid.UnaryHybridCF.calculate(UnaryHybridCF.java:61)
	at net.imagej.ops.special.hybrid.UnaryHybridCF.run(UnaryHybridCF.java:71)
	at net.imagej.ops.special.hybrid.UnaryHybridCFI.run(UnaryHybridCFI.java:66)
	at net.imagej.ops.special.hybrid.UnaryHybridCFI.run(UnaryHybridCFI.java:80)
	at org.scijava.command.CommandModule.run(CommandModule.java:196)
	at net.imagej.ops.OpEnvironment.run(OpEnvironment.java:960)
	at net.imagej.ops.OpEnvironment.run(OpEnvironment.java:157)
	at net.imagej.ops.math.MathNamespace.add(MathNamespace.java:257)
	at net.imagej.ops.math.MathNamespace$add.call(Unknown Source)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:47)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:125)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:148)
	at New_.run(New_.groovy:6)
	at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:317)
	at org.codehaus.groovy.jsr223.GroovyScriptEngineImpl.eval(GroovyScriptEngineImpl.java:155)
	at org.scijava.plugins.scripting.groovy.GroovyScriptLanguage$1.eval(GroovyScriptLanguage.java:97)
	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
	at org.scijava.script.ScriptModule.run(ScriptModule.java:164)
	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:163)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:124)
	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:63)
	at org.scijava.thread.DefaultThreadService.lambda$wrap$2(DefaultThreadService.java:225)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)

```
